### PR TITLE
add resources build_folder & build_folder_permissions

### PR DESCRIPTION
**Pull request by bobmhong**
## All Submissions:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.
* [x] All new and existing tests passed.
* [x] My code follows the code style of this project.
* [x] I ran lint checks locally prior to submission. (note: when running make ci locally, I noticed linting errors in the website docs, but not on any files I added)
* [x] Have you checked to ensure there aren't other open PRs for the same update/change?

## What about the current behavior has changed?

Implement resources azuredevops_build_folder and azuredevops_build_folder_permissions. 

Issue Number:
https://github.com/microsoft/terraform-provider-azuredevops/issues/353

## Does this introduce a change to `go.mod`, `go.sum` or `vendor/`?

- [ ] Yes
- [x] No

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Any relevant logs, error output, etc?

https://pst.klgrth.io/paste/umbwg

### CODE REVIEW

**Comment by ghost** [comment]
[CLA assistant check](https://cla.opensource.microsoft.com/microsoft/terraform-provider-azuredevops?pullRequest=604) <br/>All CLA requirements met.

File: azuredevops/internal/acceptancetests/resource_folder_test.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+package acceptancetests
+
+import (
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/acceptancetests/testutils"
+)
+
+func TestBuildFolder(t *testing.T) {
```
Comment (xuzhang3) [inline azuredevops/internal/acceptancetests/resource_folder_test.go]: Acctest function should named with `TestAcc` prefix

File: azuredevops/internal/service/build/resource_build_folder.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
+package build
+
+import (
+	"fmt"
+	"log"
+
+	// "strconv"
+	"strings"
```
Comment (xuzhang3) [inline azuredevops/internal/service/build/resource_build_folder.go]:
```suggestion
	"fmt"
	"log"
	"strings"
```

File: azuredevops/internal/service/build/resource_build_folder.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+	"github.com/microsoft/azure-devops-go-api/azuredevops/v6/build"
+	"github.com/microsoft/azure-devops-go-api/azuredevops/v6/core"
+	"github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/client"
+	"github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/utils"
+	"github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/utils/converter"
+	"github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/utils/tfhelper"
+	"github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/utils/validate"
+)
+
+// ResourceBuildDefinition schema and implementation for build folder resource
```
Comment (xuzhang3) [inline azuredevops/internal/service/build/resource_build_folder.go]:
```suggestion
// ResourceBuildFolder schema and implementation for build folder resource
```

File: azuredevops/internal/service/build/resource_build_folder.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+func resourceBuildFolderCreate(d *schema.ResourceData, m interface{}) error {
+	clients := m.(*client.AggregatedClient)
+
+	projectID := d.Get("project_id").(string)
+	description := d.Get("description").(string)
+	path := d.Get("path").(string)
+
+	createdBuildFolder, err := createBuildFolder(clients, path, projectID, description)
+	if err != nil {
+		return fmt.Errorf("error creating resource Build Folder: %+v", err)
```
Comment (xuzhang3) [inline azuredevops/internal/service/build/resource_build_folder.go]:
```suggestion
				return fmt.Errorf(" failed creating resource Build Folder, project ID: %s. Error: %+v", projectID, err)

```

File: azuredevops/internal/service/build/resource_build_folder.go [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+	return nil
+}
+
+func resourceBuildFolderUpdate(d *schema.ResourceData, m interface{}) error {
+	clients := m.(*client.AggregatedClient)
+
+	path := d.Get("path").(string)
+	buildFolder, projectID, err := expandBuildFolder(d)
+	if err != nil {
+		return err
```
Comment (xuzhang3) [inline azuredevops/internal/service/build/resource_build_folder.go]:
```suggestion
		return fmt.Errorf(" failed to expand build folder configurations. Project ID: %s , Error: %+v", projectID, err)
```

File: azuredevops/internal/service/build/resource_build_folder.go [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+	}
+
+	updatedBuildFolder, err := clients.BuildClient.UpdateFolder(m.(*client.AggregatedClient).Ctx, build.UpdateFolderArgs{
+		Project: &projectID,
+		Path:    &path,
+		Folder:  buildFolder,
+	})
+
+	if err != nil {
+		return err
```
Comment (xuzhang3) [inline azuredevops/internal/service/build/resource_build_folder.go]:
```suggestion
		return fmt.Errorf("failed to update build folder.  Project ID: %s, Error: %+v ", projectID, err)
```

File: azuredevops/internal/service/build/resource_build_folder.go [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+		Description: &description,
+		Path:        &path,
+		Project:     &projectReference,
+	}
+
+	createdBuild, err := clients.BuildClient.CreateFolder(clients.Ctx, build.CreateFolderArgs{
+		Folder:  &folder,
+		Project: &project,
+		Path:    &path,
+	})
```
Comment (xuzhang3) [inline azuredevops/internal/service/build/resource_build_folder.go]:
```suggestion
	createdBuild, err := clients.BuildClient.CreateFolder(clients.Ctx, build.CreateFolderArgs{
		Folder: &build.Folder{
			Description: &description,
			Path:        &path,
			Project: &core.TeamProjectReference{
				Id: &projectUuid,
			},
		},
		Project: &project,
		Path:    &path,
	})
```

File: azuredevops/internal/service/permissions/resource_build_folder_permissions.go [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+		return "", fmt.Errorf("Failed to get 'path' from schema")
+	}
+
+	buildFolders, err := clients.BuildClient.GetFolders(clients.Ctx, build.GetFoldersArgs{
+		Project: converter.String(projectID.(string)),
+		Path:    converter.String(buildFolderPath.(string)),
+	})
+
+	if err != nil {
+		return "", err
```
Comment (xuzhang3) [inline azuredevops/internal/service/permissions/resource_build_folder_permissions.go]:
```suggestion
		return "", fmt.Errorf(" faild to get the folder. Project ID: %s, path: %s. %+v", projectID, buildFolderPath, err)
```

File: main.go [hunk 9]
```diff
@@ -1,13 +1,21 @@
...
 )
 
 func main() {
+	var debug bool
+
+	flag.BoolVar(&debug, "debug", false, "set to true to run the provider with support for debuggers like delve")
+	flag.Parse()
+
 	plugin.Serve(&plugin.ServeOpts{
+		Debug: debug,
```
Comment (xuzhang3) [inline main.go]:
```suggestion
		Debug:        debug,
		ProviderAddr: "registry.terraform.io/microsoft/azuredevops",
```

File: website/docs/r/build_folder_permissions.html.markdown [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+  name       = "Readers"
+}
+
+resource "azuredevops_build_folder" "example" {
+  project_id  = azuredevops_project.example.id
+  path        = "\\ExampleFolder"
+  description = "ExampleFolder description"
+}
+
+resource "azuredevops_build_folder_permissions" "example" {
```
Comment (xuzhang3) [inline website/docs/r/build_folder_permissions.html.markdown:36]: The example `azuredevops_build_folder_permissions.example`  missing the path configuration `path        = azuredevops_build_folder.example.path`

**Review by xuzhang3 — CHANGES_REQUESTED** [review]

**Comment by bobmhong** [comment]
@xuzhang3 Thanks for the review - I've added a commit with your suggested changes.

**Review by xuzhang3 — CHANGES_REQUESTED** [review]
@bobmhong   a few changes requested, others LGTM.
1. `resource_folder_test.go` should rename to `resource_build_folder_test`
2. `website/docs/r/build_folder.html.markdown` and `website/docs/r/build_folder_permissions.html.markdown` should add to `azuredevops.erb`

**Comment by xuzhang3** [comment]
@bobmhong LGTM  🚢

**Review by xuzhang3 — APPROVED** [review]

**Pull request merged**